### ISSUE TITLE
bug: "header not found" error when using client.readContract

### ISSUE THREAD

**Issue by danconnell** [ponder-sh/ponder#1189]
Hey, I've been trying to upgrade from 0.5 to 0.6 but have been unable to do so because we are getting `header not found` errors when using `client.readContract`.

Here is an example (RPC URL sanitized):

```
DEFAULT 2024-10-25T16:53:45.981973Z 4:53:45 PM WARN sync Failed 'eth_call' RPC request
DEFAULT 2024-10-25T16:53:46.023904Z 4:53:45 PM ERROR indexing Error while processing 'NFTAttributesManagerV1:AttributesUpdated' event in 'blast' block 10532705
DEFAULT 2024-10-25T16:53:46.023953Z ContractFunctionExecutionError: Missing or invalid parameters.
DEFAULT 2024-10-25T16:53:46.023959Z Double check you have provided the correct parameters.
DEFAULT 2024-10-25T16:53:46.023968Z URL: https://blastl2-mainnet.blastapi.io/xxx
DEFAULT 2024-10-25T16:53:46.023974Z Request body: {"method":"eth_call","params":[{"data":"0x4378a6e30000000000000000000000000000000000000000000000000000000000000dd5","to":"0xEC86088Bf8dACa12d60146DFB66bCaAf679E3dbF"},"0xa0b761"]}
DEFAULT 2024-10-25T16:53:46.023982Z
DEFAULT 2024-10-25T16:53:46.023988Z Raw Call Arguments:
DEFAULT 2024-10-25T16:53:46.023996Z to: 0xEC86088Bf8dACa12d60146DFB66bCaAf679E3dbF
DEFAULT 2024-10-25T16:53:46.024003Z data: 0x4378a6e30000000000000000000000000000000000000000000000000000000000000dd5
DEFAULT 2024-10-25T16:53:46.024009Z
DEFAULT 2024-10-25T16:53:46.024015Z Contract Call:
DEFAULT 2024-10-25T16:53:46.024022Z address: 0xEC86088Bf8dACa12d60146DFB66bCaAf679E3dbF
DEFAULT 2024-10-25T16:53:46.024030Z function: getAttributes(uint256 _tokenId)
DEFAULT 2024-10-25T16:53:46.024036Z args: (3541)
DEFAULT 2024-10-25T16:53:46.024044Z Docs: https://viem.sh/docs/contract/readContract
DEFAULT 2024-10-25T16:53:46.024051Z Details: header not found
DEFAULT 2024-10-25T16:53:46.024058Z Version: viem@2.21.35
DEFAULT 2024-10-25T16:53:46.024064Z at /usr/src/app/src/handlers/nft-attributes-manager-v1.ts:12:9
DEFAULT 2024-10-25T16:53:46.024071Z 10 | const coreAttributes = skipRPCCalls
DEFAULT 2024-10-25T16:53:46.024077Z 11 | ? {}
DEFAULT 2024-10-25T16:53:46.024092Z > 12 | : await context.client.readContract({
DEFAULT 2024-10-25T16:53:46.024099Z | ^
DEFAULT 2024-10-25T16:53:46.024106Z 13 | abi: NFTAttributesManagerV1.abi,
DEFAULT 2024-10-25T16:53:46.024112Z 14 | address: NFTAttributesManagerV1.address!,
DEFAULT 2024-10-25T16:53:46.024119Z 15 | functionName: "getAttributes",
DEFAULT 2024-10-25T16:53:46.024156Z caused by: CallExecutionError: Missing or invalid parameters.
DEFAULT 2024-10-25T16:53:46.024163Z Double check you have provided the correct parameters.
DEFAULT 2024-10-25T16:53:46.024175Z URL: https://blastl2-mainnet.blastapi.io/xxx
DEFAULT 2024-10-25T16:53:46.024182Z Request body: {"method":"eth_call","params":[{"data":"0x4378a6e30000000000000000000000000000000000000000000000000000000000000dd5","to":"0xEC86088Bf8dACa12d60146DFB66bCaAf679E3dbF"},"0xa0b761"]}
DEFAULT 2024-10-25T16:53:46.024188Z
DEFAULT 2024-10-25T16:53:46.024194Z Raw Call Arguments:
DEFAULT 2024-10-25T16:53:46.024200Z to: 0xEC86088Bf8dACa12d60146DFB66bCaAf679E3dbF
DEFAULT 2024-10-25T16:53:46.024216Z data: 0x4378a6e30000000000000000000000000000000000000000000000000000000000000dd5
DEFAULT 2024-10-25T16:53:46.024223Z Details: header not found
DEFAULT 2024-10-25T16:53:46.024231Z Version: viem@2.21.35
DEFAULT 2024-10-25T16:53:46.024242Z at getCallError (file:///usr/src/app/node_modules/.pnpm/viem@2.21.35_typescript@5.4.4/node_modules/viem/utils/errors/getCallError.ts:42:10)
DEFAULT 2024-10-25T16:53:46.024249Z at call (file:///usr/src/app/node_modules/.pnpm/viem@2.21.35_typescript@5.4.4/node_modules/viem/actions/public/call.ts:282:11)
DEFAULT 2024-10-25T16:53:46.024255Z at processTicksAndRejections (node:internal/process/task_queues:95:5)
DEFAULT 2024-10-25T16:53:46.024261Z at readContract (file:///usr/src/app/node_modules/.pnpm/viem@2.21.35_typescript@5.4.4/node_modules/viem/actions/public/readContract.ts:120:22)
DEFAULT 2024-10-25T16:53:46.024267Z at /usr/src/app/src/handlers/nft-attributes-manager-v1.ts:12:9
DEFAULT 2024-10-25T16:53:46.024274Z at executeLog (file:///usr/src/app/node_modules/.pnpm/@ponder+core@0.6.19_@types+node@20.12.6_hono@4.5.1_typescript@5.4.4_viem@2.21.35/node_modules/@ponder/core/src/indexing/service.ts:509:5)
DEFAULT 2024-10-25T16:53:46.024280Z at processEvents (file:///usr/src/app/node_modules/.pnpm/@ponder+core@0.6.19_@types+node@20.12.6_hono@4.5.1_typescript@5.4.4_viem@2.21.35/node_modules/@ponder/core/src/indexing/service.ts:292:24)
DEFAULT 2024-10-25T16:53:46.024287Z at handleEvents (file:///usr/src/app/node_modules/.pnpm/@ponder+core@0.6.19_@types+node@20.12.6_hono@4.5.1_typescript@5.4.4_viem@2.21.35/node_modules/@ponder/core/src/bin/utils/run.ts:95:12)
DEFAULT 2024-10-25T16:53:46.024294Z at worker (file:///usr/src/app/node_modules/.pnpm/@ponder+core@0.6.19_@types+node@20.12.6_hono@4.5.1_typescript@5.4.4_viem@2.21.35/node_modules/@ponder/core/src/bin/utils/run.ts:105:26)
DEFAULT 2024-10-25T16:53:46.024304Z caused by: InvalidInputRpcError: Missing or invalid parameters.
DEFAULT 2024-10-25T16:53:46.024310Z Double check you have provided the correct parameters.
DEFAULT 2024-10-25T16:53:46.024317Z URL: https://blastl2-mainnet.blastapi.io/xxx
DEFAULT 2024-10-25T16:53:46.024323Z Request body: {"method":"eth_call","params":[{"data":"0x4378a6e30000000000000000000000000000000000000000000000000000000000000dd5","to":"0xEC86088Bf8dACa12d60146DFB66bCaAf679E3dbF"},"0xa0b761"]}
DEFAULT 2024-10-25T16:53:46.024330Z Details: header not found
DEFAULT 2024-10-25T16:53:46.024336Z Version: viem@2.21.35
DEFAULT 2024-10-25T16:53:46.024343Z at withRetry.delay.count.count (file:///usr/src/app/node_modules/.pnpm/viem@2.21.35_typescript@5.4.4/node_modules/viem/utils/buildRequest.ts:141:25)
DEFAULT 2024-10-25T16:53:46.024349Z at processTicksAndRejections (node:internal/process/task_queues:95:5)
DEFAULT 2024-10-25T16:53:46.024356Z at attemptRetry (file:///usr/src/app/node_modules/.pnpm/viem@2.21.35_typescript@5.4.4/node_modules/viem/utils/promise/withRetry.ts:44:22)
DEFAULT 2024-10-25T16:53:46.024365Z caused by: InvalidInputRpcError: Missing or invalid parameters.
DEFAULT 2024-10-25T16:53:46.024371Z Double check you have provided the correct parameters.
DEFAULT 2024-10-25T16:53:46.024378Z URL: https://blastl2-mainnet.blastapi.io/xxx
DEFAULT 2024-10-25T16:53:46.024384Z Request body: {"method":"eth_call","params":[{"data":"0x4378a6e30000000000000000000000000000000000000000000000000000000000000dd5","to":"0xEC86088Bf8dACa12d60146DFB66bCaAf679E3dbF"},"0xa0b761"]}
DEFAULT 2024-10-25T16:53:46.024390Z Details: header not found
DEFAULT 2024-10-25T16:53:46.024397Z Version: viem@2.21.35
DEFAULT 2024-10-25T16:53:46.024403Z at withRetry.delay.count.count (file:///usr/src/app/node_modules/.pnpm/viem@2.21.35_typescript@5.4.4/node_modules/viem/utils/buildRequest.ts:141:25)
DEFAULT 2024-10-25T16:53:46.024410Z at processTicksAndRejections (node:internal/process/task_queues:95:5)
DEFAULT 2024-10-25T16:53:46.024416Z at attemptRetry (file:///usr/src/app/node_modules/.pnpm/viem@2.21.35_typescript@5.4.4/node_modules/viem/utils/promise/withRetry.ts:44:22)
DEFAULT 2024-10-25T16:53:46.024429Z caused by: RpcRequestError: RPC Request failed.
DEFAULT 2024-10-25T16:53:46.024436Z URL: https://blastl2-mainnet.blastapi.io/xxx
DEFAULT 2024-10-25T16:53:46.024443Z Request body: {"method":"eth_call","params":[{"data":"0x4378a6e30000000000000000000000000000000000000000000000000000000000000dd5","to":"0xEC86088Bf8dACa12d60146DFB66bCaAf679E3dbF"},"0xa0b761"]}
DEFAULT 2024-10-25T16:53:46.024450Z Details: header not found
DEFAULT 2024-10-25T16:53:46.024456Z Version: viem@2.21.35
DEFAULT 2024-10-25T16:53:46.024462Z at request (file:///usr/src/app/node_modules/.pnpm/viem@2.21.35_typescript@5.4.4/node_modules/viem/clients/transports/http.ts:132:19)
DEFAULT 2024-10-25T16:53:46.024468Z at processTicksAndRejections (node:internal/process/task_queues:95:5)
DEFAULT 2024-10-25T16:53:46.024473Z at withRetry.delay.count.count (file:///usr/src/app/node_modules/.pnpm/viem@2.21.35_typescript@5.4.4/node_modules/viem/utils/buildRequest.ts:118:22)
DEFAULT 2024-10-25T16:53:46.024480Z at attemptRetry (file:///usr/src/app/node_modules/.pnpm/viem@2.21.35_typescript@5.4.4/node_modules/viem/utils/promise/withRetry.ts:44:22)
DEFAULT 2024-10-25T16:53:46.024486Z caused by:
DEFAULT 2024-10-25T16:53:46.024492Z Event arguments:
DEFAULT 2024-10-25T16:53:46.024498Z _tokenId 3541
DEFAULT 2024-10-25T16:53:46.024504Z 4:53:46 PM WARN process Encountered indexing error, starting shutdown sequence
DEFAULT 2024-10-25T16:53:46.103130Z 4:53:46 PM FATAL process Finished shutdown sequence, terminating (exit code 1)
WARNING 2024-10-25T16:53:47.055704503Z Container called exit(1).
```

This error is very intermittent and we make tens of thousands of these calls when indexing and they all are fine and then all of a sudden it errors hours later. There are no logs of a fork at this time (I have seen logs for forks at other times where we get this error and then it recovers).

I've tried setting `disableCache: true` and that simply caused me to get the same error on a much older block.

This error does not occur on 0.5.3 but occurs within a few hours of a deployment of 0.6 (have tried many versions including the latest 0.6.19).

**Comment by kyscott18** [comment]
Thanks for opening. My first intuition when seeing an error like this is that the rpc is not an archive node. It is suspicious that it's not reproducible and the behavior changed between ponder versions. Have you tried this with a different rpc provider?

I don't think `disableCache` will have any effect.

**Comment by danconnell** [comment]
I haven't tried a different RPC provider as I don't have another paid one to use and for the number of events we're indexing we need a paid one. But the one we're using does say (on https://blastapi.io/dedicated-endpoints):

```
Most of Blast's dedicated endpoints are powered by archive nodes, giving you access to all the historical data you need to build the next generation of Web3 products.
```

This is occurring during realtime indexing as well so I don't think it's because it's not an archive node. Here's another example showing it during realtime:

```
DEFAULT 2024-10-25T09:51:07.845973Z 9:51:07 AM INFO realtime Synced 2 logs from 'blast' block 10520026
DEFAULT 2024-10-25T09:51:07.984536Z 9:51:07 AM WARN sync Failed 'eth_call' RPC request
DEFAULT 2024-10-25T09:51:08.028514Z 9:51:08 AM ERROR indexing Error while processing 'NFTAttributesManagerV1:AttributesUpdated' event in 'blast' block 10520026
DEFAULT 2024-10-25T09:51:08.028559Z ContractFunctionExecutionError: Missing or invalid parameters.
DEFAULT 2024-10-25T09:51:08.028563Z Double check you have provided the correct parameters.
DEFAULT 2024-10-25T09:51:08.028571Z URL: https://blastl2-mainnet.blastapi.io/xxx
DEFAULT 2024-10-25T09:51:08.028576Z Request body: {"method":"eth_call","params":[{"data":"0x4378a6e30000000000000000000000000000000000000000000000000000000000000ce2","to":"0xEC86088Bf8dACa12d60146DFB66bCaAf679E3dbF"},"0xa085da"]}
DEFAULT 2024-10-25T09:51:08.028583Z
DEFAULT 2024-10-25T09:51:08.028590Z Raw Call Arguments:
DEFAULT 2024-10-25T09:51:08.028595Z to: 0xEC86088Bf8dACa12d60146DFB66bCaAf679E3dbF
DEFAULT 2024-10-25T09:51:08.028601Z data: 0x4378a6e30000000000000000000000000000000000000000000000000000000000000ce2
DEFAULT 2024-10-25T09:51:08.028605Z
DEFAULT 2024-10-25T09:51:08.028611Z Contract Call:
DEFAULT 2024-10-25T09:51:08.028615Z address: 0xEC86088Bf8dACa12d60146DFB66bCaAf679E3dbF
DEFAULT 2024-10-25T09:51:08.028620Z function: getAttributes(uint256 _tokenId)
DEFAULT 2024-10-25T09:51:08.028625Z args: (3298)
DEFAULT 2024-10-25T09:51:08.028631Z Docs: https://viem.sh/docs/contract/readContract
DEFAULT 2024-10-25T09:51:08.028637Z Details: header not found
DEFAULT 2024-10-25T09:51:08.028641Z Version: viem@2.21.32
DEFAULT 2024-10-25T09:51:08.028647Z at /usr/src/app/src/handlers/nft-attributes-manager-v1.ts:12:9
DEFAULT 2024-10-25T09:51:08.028652Z 10 | const coreAttributes = skipRPCCalls
DEFAULT 2024-10-25T09:51:08.028657Z 11 | ? {}
DEFAULT 2024-10-25T09:51:08.028662Z > 12 | : await context.client.readContract({
DEFAULT 2024-10-25T09:51:08.028666Z | ^
DEFAULT 2024-10-25T09:51:08.028671Z 13 | abi: NFTAttributesManagerV1.abi,
DEFAULT 2024-10-25T09:51:08.028677Z 14 | address: NFTAttributesManagerV1.address!,
DEFAULT 2024-10-25T09:51:08.028682Z 15 | functionName: "getAttributes",
DEFAULT 2024-10-25T09:51:08.028691Z caused by: CallExecutionError: Missing or invalid parameters.
DEFAULT 2024-10-25T09:51:08.028695Z Double check you have provided the correct parameters.
DEFAULT 2024-10-25T09:51:08.028702Z URL: https://blastl2-mainnet.blastapi.io/xxx
DEFAULT 2024-10-25T09:51:08.028707Z Request body: {"method":"eth_call","params":[{"data":"0x4378a6e30000000000000000000000000000000000000000000000000000000000000ce2","to":"0xEC86088Bf8dACa12d60146DFB66bCaAf679E3dbF"},"0xa085da"]}
DEFAULT 2024-10-25T09:51:08.028712Z
DEFAULT 2024-10-25T09:51:08.028716Z Raw Call Arguments:
DEFAULT 2024-10-25T09:51:08.028721Z to: 0xEC86088Bf8dACa12d60146DFB66bCaAf679E3dbF
DEFAULT 2024-10-25T09:51:08.028727Z data: 0x4378a6e30000000000000000000000000000000000000000000000000000000000000ce2
DEFAULT 2024-10-25T09:51:08.028732Z Details: header not found
DEFAULT 2024-10-25T09:51:08.028736Z Version: viem@2.21.32
DEFAULT 2024-10-25T09:51:08.028748Z at getCallError (file:///usr/src/app/node_modules/.pnpm/viem@2.21.32_typescript@5.4.4/node_modules/viem/utils/errors/getCallError.ts:42:10)
DEFAULT 2024-10-25T09:51:08.028754Z at call (file:///usr/src/app/node_modules/.pnpm/viem@2.21.32_typescript@5.4.4/node_modules/viem/actions/public/call.ts:282:11)
DEFAULT 2024-10-25T09:51:08.028759Z at processTicksAndRejections (node:internal/process/task_queues:95:5)
DEFAULT 2024-10-25T09:51:08.028765Z at readContract (file:///usr/src/app/node_modules/.pnpm/viem@2.21.32_typescript@5.4.4/node_modules/viem/actions/public/readContract.ts:120:22)
DEFAULT 2024-10-25T09:51:08.028769Z at /usr/src/app/src/handlers/nft-attributes-manager-v1.ts:12:9
DEFAULT 2024-10-25T09:51:08.028774Z at executeLog (file:///usr/src/app/node_modules/.pnpm/@ponder+core@0.6.18_@types+node@20.12.6_hono@4.5.1_typescript@5.4.4_viem@2.21.32/node_modules/@ponder/core/src/indexing/service.ts:509:5)
DEFAULT 2024-10-25T09:51:08.028779Z at processEvents (file:///usr/src/app/node_modules/.pnpm/@ponder+core@0.6.18_@types+node@20.12.6_hono@4.5.1_typescript@5.4.4_viem@2.21.32/node_modules/@ponder/core/src/indexing/service.ts:292:24)
DEFAULT 2024-10-25T09:51:08.028784Z at handleEvents (file:///usr/src/app/node_modules/.pnpm/@ponder+core@0.6.18_@types+node@20.12.6_hono@4.5.1_typescript@5.4.4_viem@2.21.32/node_modules/@ponder/core/src/bin/utils/run.ts:95:12)
DEFAULT 2024-10-25T09:51:08.028791Z at worker (file:///usr/src/app/node_modules/.pnpm/@ponder+core@0.6.18_@types+node@20.12.6_hono@4.5.1_typescript@5.4.4_viem@2.21.32/node_modules/@ponder/core/src/bin/utils/run.ts:105:26)
DEFAULT 2024-10-25T09:51:08.028797Z caused by: InvalidInputRpcError: Missing or invalid parameters.
DEFAULT 2024-10-25T09:51:08.028801Z Double check you have provided the correct parameters.
DEFAULT 2024-10-25T09:51:08.028807Z URL: https://blastl2-mainnet.blastapi.io/xxx
DEFAULT 2024-10-25T09:51:08.028812Z Request body: {"method":"eth_call","params":[{"data":"0x4378a6e30000000000000000000000000000000000000000000000000000000000000ce2","to":"0xEC86088Bf8dACa12d60146DFB66bCaAf679E3dbF"},"0xa085da"]}
DEFAULT 2024-10-25T09:51:08.028817Z Details: header not found
DEFAULT 2024-10-25T09:51:08.028821Z Version: viem@2.21.32
DEFAULT 2024-10-25T09:51:08.028827Z at withRetry.delay.count.count (file:///usr/src/app/node_modules/.pnpm/viem@2.21.32_typescript@5.4.4/node_modules/viem/utils/buildRequest.ts:141:25)
DEFAULT 2024-10-25T09:51:08.028832Z at processTicksAndRejections (node:internal/process/task_queues:95:5)
DEFAULT 2024-10-25T09:51:08.028838Z at attemptRetry (file:///usr/src/app/node_modules/.pnpm/viem@2.21.32_typescript@5.4.4/node_modules/viem/utils/promise/withRetry.ts:44:22)
DEFAULT 2024-10-25T09:51:08.028845Z caused by: InvalidInputRpcError: Missing or invalid parameters.
DEFAULT 2024-10-25T09:51:08.028850Z Double check you have provided the correct parameters.
DEFAULT 2024-10-25T09:51:08.028856Z URL: https://blastl2-mainnet.blastapi.io/xxx
DEFAULT 2024-10-25T09:51:08.028861Z Request body: {"method":"eth_call","params":[{"data":"0x4378a6e30000000000000000000000000000000000000000000000000000000000000ce2","to":"0xEC86088Bf8dACa12d60146DFB66bCaAf679E3dbF"},"0xa085da"]}
DEFAULT 2024-10-25T09:51:08.028867Z Details: header not found
DEFAULT 2024-10-25T09:51:08.028871Z Version: viem@2.21.32
DEFAULT 2024-10-25T09:51:08.028876Z at withRetry.delay.count.count (file:///usr/src/app/node_modules/.pnpm/viem@2.21.32_typescript@5.4.4/node_modules/viem/utils/buildRequest.ts:141:25)
DEFAULT 2024-10-25T09:51:08.028880Z at processTicksAndRejections (node:internal/process/task_queues:95:5)
DEFAULT 2024-10-25T09:51:08.028885Z at attemptRetry (file:///usr/src/app/node_modules/.pnpm/viem@2.21.32_typescript@5.4.4/node_modules/viem/utils/promise/withRetry.ts:44:22)
DEFAULT 2024-10-25T09:51:08.028891Z caused by: RpcRequestError: RPC Request failed.
DEFAULT 2024-10-25T09:51:08.028897Z URL: https://blastl2-mainnet.blastapi.io/xxx
DEFAULT 2024-10-25T09:51:08.028902Z Request body: {"method":"eth_call","params":[{"data":"0x4378a6e30000000000000000000000000000000000000000000000000000000000000ce2","to":"0xEC86088Bf8dACa12d60146DFB66bCaAf679E3dbF"},"0xa085da"]}
DEFAULT 2024-10-25T09:51:08.028907Z Details: header not found
DEFAULT 2024-10-25T09:51:08.028911Z Version: viem@2.21.32
DEFAULT 2024-10-25T09:51:08.028916Z at request (file:///usr/src/app/node_modules/.pnpm/viem@2.21.32_typescript@5.4.4/node_modules/viem/clients/transports/http.ts:132:19)
DEFAULT 2024-10-25T09:51:08.028921Z at processTicksAndRejections (node:internal/process/task_queues:95:5)
DEFAULT 2024-10-25T09:51:08.028926Z at withRetry.delay.count.count (file:///usr/src/app/node_modules/.pnpm/viem@2.21.32_typescript@5.4.4/node_modules/viem/utils/buildRequest.ts:118:22)
DEFAULT 2024-10-25T09:51:08.028932Z at attemptRetry (file:///usr/src/app/node_modules/.pnpm/viem@2.21.32_typescript@5.4.4/node_modules/viem/utils/promise/withRetry.ts:44:22)
DEFAULT 2024-10-25T09:51:08.028936Z caused by:
DEFAULT 2024-10-25T09:51:08.028941Z Event arguments:
DEFAULT 2024-10-25T09:51:08.028945Z _tokenId 3298
DEFAULT 2024-10-25T09:51:08.028950Z 9:51:08 AM WARN process Encountered indexing error, starting shutdown sequence
DEFAULT 2024-10-25T09:51:08.036163Z 9:51:08 AM FATAL process Finished shutdown sequence, terminating (exit code 1)
WARNING 2024-10-25T09:51:08.879983184Z Container called exit(1).
```

**Comment by kyscott18** [comment]
Just released v0.6.21 that should more aggressively retry this error. Feel free to reopen this if you are still having issues.
